imap/guesstz.c

```diff
@@ -0,0 +1,914 @@
+            r = cyrusdb_store(gtzdb->db, buf_base(key), buf_len(key),
+                    buf_base(data), buf_len(data), &gtzdb->tid);
+            if (r) goto done;
+            /* Reset state */
+            prevoffset = offsetrec->offset;
+            buf_reset(key);
+            buf_reset(data);
+        }
+        buf_appendbit32(data, offsetrec->tznum);
+    }
+    if (buf_len(data)) {
+        /* Write previous record */
+        buf_putc(key, 'O');
+        buf_appendbit32(key, prevoffset);
+        r = cyrusdb_store(gtzdb->db, buf_base(key), buf_len(key),
+                buf_base(data), buf_len(data), &gtzdb->tid);
+        if (r) goto done;
+    }
+
+done:
+    buf_reset(key);
+    buf_reset(data);
+    return r;
+}
+
+static int store_config(struct guesstzdb *gtzdb)
+{
+    struct buf *key = &gtzdb->buf1;
+    buf_reset(key);
+    buf_putc(key, 'C');
+
+    struct buf *data = &gtzdb->buf2;
+    buf_reset(data);
+
+    /* 1-byte version */
+    buf_appendmap(data, &gtzdb->version, 1);
+
+    /* 8-byte timestamp of creation time */
+    int64_t t = gtzdb->created;
+    buf_appendmap(data, (char *)&t, 8);
+
+    /* 16-byte time range */
+    const icaltimezone *utc = icaltimezone_get_utc_timezone();
+    int64_t tstart = icaltime_as_timet_with_zone(gtzdb->dbstart, utc);
+    int64_t tend = icaltime_as_timet_with_zone(gtzdb->dbend, utc);
+    buf_appendmap(data, (char *)&tstart, 8);
+    buf_appendmap(data, (char *)&tend, 8);
+
+    /* Zero-terminated IANA version name */
+    buf_appendcstr(data, gtzdb->ianaversion);
+    buf_putc(data, '\0');
+
+    int r = cyrusdb_store(gtzdb->db, buf_base(key), buf_len(key),
+            buf_base(data), buf_len(data), &gtzdb->tid);
+
+    buf_free(key);
+    buf_free(data);
+    return r;
+}
+
+static char *read_icalstream(char *s, size_t size, void *d)
+{
+    return fgets(s, (int)size, (FILE*)d);
+}
+
+static int create_from_zonedir(struct guesstzdb *gtzdb,
+                               const char *zoneinfo_dir,
+                               icaltimetype start, icaltimetype end)
+{
+    dynarray_t offsetrecs;
+    dynarray_init(&offsetrecs, sizeof(struct offsetrec));
+    arrayu64_t uniqoffsets = ARRAYU64_INITIALIZER;
+    dynarray_t obsrecs;
+    dynarray_init(&obsrecs, sizeof(struct obsrec));
+    int r = 0;
+
+    /* Traverse timezone definitions */
+    uint32_t tznum = 1; // time zone numeric ids start at 1
+    FTS *fts = NULL;
+
+    char *paths[2] = { (char *) zoneinfo_dir, NULL };
+    fts = fts_open(paths, 0, NULL);
+    if (!fts) {
+        xsyslog(LOG_ERR, "fts_open", "zoneinfo_dir=<%s> err=<%s>",
+                zoneinfo_dir, strerror(errno));
+        r = CYRUSDB_IOERROR;
+        goto done;
+    }
+
+    FTSENT *fe;
+    while ((fe = fts_read(fts))) {
+        if (fe->fts_info != FTS_F) {
+            continue;
+        }
+
+        FILE *stream = fopen(fe->fts_path, "r");
+        if (!stream) continue;
+
+        icalparser *parser = icalparser_new();
```

Comment:
Wouldn't it be easier to just map_refresh() the file and use icalparser_parse_string()?

Reply:
done